### PR TITLE
Add a less used worker choice strategy

### PR DESCRIPTION
**Pull request by jerome-benoit**
And cleanup the implementation of worker selection strategies.

Signed-off-by: Jérôme Benoit <jerome.benoit@sap.com>

## PR Checklist

- [x] Please add a description of your changes.
- [ ] We need your changes to come with unit tests in order to keep this project in quality and easy to maintain.
- [ ] Please add a link to the open issue or task that this pull request will resolve.
- [ ] Add a note to the changelog to indicate the change.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

If yes please indicate it and migration info into the CHANGELOG.md file.

---

### CODE REVIEW

**Review by pioardi — COMMENTED** [review]
I like it, really good.
We should only add more documentation on the new option and strategies.
Let me know of you sant ti add doc later than I agree to proceed

**Comment by jerome-benoit** [comment]
> I like it, really good.
> We should only add more documentation on the new option and strategies.
> Let me know of you sant ti add doc later than I agree to proceed. 

The checks are failing because of the level of coverage. 
For me, the @Shinigami92 work and mine (3 stacked PRs) over his are in a mergeable state even if coverage is lower. They are still probably some small code cleanups to do but it installs some fondation regarding dynamic pluggable behaviour on pools without regression. 
I think cleanups, unit tests, benchmarks and documentation can be done later on incrementally. The coverage level will probably need to lowered until then ...

**Comment by Shinigami92** [comment]
I will look into all of this tomorrow
Good night :wave:

**Comment by pioardi** [comment]
> For me, the @Shinigami92 work and mine (3 stacked PRs) over his are in a mergeable state even if coverage is lower. They are still probably some small code cleanups to do but it installs some fondation regarding dynamically pluggable behaviours on pool without regression.

For me is ok to merge on other branches if coverage is not high, but we need to meet the criteria that are defined for coverage and quality before to merge on master branch

File: src/pools/selection-strategies.ts
Comment (pioardi) [inline src/pools/selection-strategies.ts:82]: I think is better to return directly instead of store the value into a variable and then break the cycle

File: src/pools/selection-strategies.ts
Comment (pioardi) [inline src/pools/selection-strategies.ts:86]: same as above

**Review by pioardi — COMMENTED** [review]
General shape is good, one small suggestion

File: src/pools/selection-strategies.ts
Comment (jerome-benoit) [reply src/pools/selection-strategies.ts:82]: Here I can return.

File: src/pools/selection-strategies.ts
Comment (jerome-benoit) [reply src/pools/selection-strategies.ts:86]: But here I can't, the loop need to iterate over all the map to elect the less recently used.

Comment (pioardi) [reply src/pools/selection-strategies.ts:86]: Make sense, it's ok

**Comment by Shinigami92** [comment]
@jerome-benoit Ok I will start to deep dive into your PR stacks and will maybe add some code to it
So I will let you know when to check it

**Comment by jerome-benoit** [comment]
> @jerome-benoit Ok I will start to deep dive into your PR stacks and will maybe add some code to it
> So I will let you know when to check it

I think that's fine to merge all of them in your PR after. They all address the same issue. 
The merge in master will require ut for coverage and documentation.

**Review by pioardi — APPROVED** [review]
Looks good

File: src/pools/selection-strategies.ts
Comment (Shinigami92) [inline src/pools/selection-strategies.ts:10]: @jerome-benoit Do you have a good description for this one?

File: src/pools/selection-strategies.ts
Comment (Shinigami92) [inline src/pools/selection-strategies.ts:14]: @jerome-benoit Do you have a good description for this one?

File: src/pools/selection-strategies.ts
Comment (Shinigami92) [inline src/pools/selection-strategies.ts:95]: This could be only produce an error if `pool.tasks` is empty :thinking:

File: src/pools/selection-strategies.ts
Comment (jerome-benoit) [reply src/pools/selection-strategies.ts:10]: Less recently used worker selection strategy ;)

File: src/pools/selection-strategies.ts
Comment (jerome-benoit) [reply src/pools/selection-strategies.ts:14]: Round robin worker selection strategy ;)

File: src/pools/selection-strategies.ts
Comment (jerome-benoit) [reply src/pools/selection-strategies.ts:95]: We can add a sanity check on the map length and then throw an error but that should not happen and I'm not in favour of adding more branches to a hot code path until a bug is found. 
I think it's more related to worker pool initialisation timing: a busy loop can be added at init to ensure we wait for at lest a worker to be created.

**Review by pioardi — APPROVED** [review]

File: src/pools/selection-strategies.ts
Comment (Shinigami92) [reply src/pools/selection-strategies.ts:95]: Maybe we should add a check for `numberOfWorkers` to be greater than or equals to `1`
Then in `AbstractPool` constructor the `createAndSetupWorker` function is always called at least once

**Comment by jerome-benoit** [comment]
I'm fine with the @Shinigami92 changes, I can't approuve my own PR but I approuve the merge on the previous PR in the stack and then the final PR.

File: src/pools/selection-strategies.ts
Comment (jerome-benoit) [reply src/pools/selection-strategies.ts:95]: Like for `checkFilePath (filePath: string) `? A 

```ts
checkNumberOfWorkers (numberOfWorkers: number) {
  if (numberOfWorkers === 0) {
    throw Error('')
  }
}
```

method will the call in the constructor? That's ok.

**Comment by Shinigami92** [comment]
@jerome-benoit Anything else you want to have changed in this PR? Otherwise I will climb down the stack-ladder down to https://github.com/pioardi/poolifier/pull/191 and merge this PR into https://github.com/pioardi/poolifier/pull/191

**Comment by jerome-benoit** [comment]
> @jerome-benoit Anything else you want to have changed in this PR? Otherwise I will climb down the stack-ladder down to #191 and merge this PR into #191

Go ahead, I'm going to add the check on numberOfWorkers in another PR right now.

**Pull request merged**
